wrappers/pkcs11/pkcs11_client.go

```diff
@@ -0,0 +1,553 @@
+		ciphertext = ciphertext[ivLength:]
+	}
+
+	if err = c.client.DecryptInit(session, []*pkcs11.Mechanism{pkcs11.NewMechanism(mechanism, iv)}, key); err != nil {
+		return nil, fmt.Errorf("failed to pkcs11 DecryptInit: %s", err)
+	}
+
+	var decrypted []byte
+	if decrypted, err = c.client.Decrypt(session, ciphertext); err != nil {
+		return nil, fmt.Errorf("failed to pkcs11 Decrypt: %s", err)
+	}
+	return decrypted, nil
+}
+
+// Create a PKCS11 client for the configured module.
+func (c *Pkcs11Client) GetClient() (*pkcs11.Ctx, error) {
+	if c.client != nil {
+		return c.client, nil
+	}
+	c.client = pkcs11.New(c.lib)
+	err := c.client.Initialize()
+	if err != nil {
+		c.client = nil
+		return nil, fmt.Errorf("failed to initialize PKCS11: %w", err)
+	}
+	return c.client, nil
+}
+
+func (c *Pkcs11Client) GetSlotForLabel() (uint, error) {
+	if c.slot != 0 {
+		return c.slot, nil
+	}
+	if c.tokenLabel == "" {
+		return 0, fmt.Errorf("not token label configured")
+	}
+	slots, _ := c.client.GetSlotList(true)
+	for _, slot := range slots {
+		tokenInfo, err := c.client.GetTokenInfo(slot)
+		if err == nil && tokenInfo.Label == c.tokenLabel {
+			c.slot = slot
+			break
+		}
+	}
+	if c.slot == 0 {
+		return 0, fmt.Errorf("failed to find token with label: %s", c.tokenLabel)
+	}
+	return c.slot, nil
+}
+
+// Open a session and perform the authentication process.
+func (c *Pkcs11Client) GetSession() (pkcs11.SessionHandle, error) {
+	if c.client == nil {
+		return 0, fmt.Errorf("PKCS11 not initialized")
+	}
+
+	if c.slot == 0 {
+		_, err := c.GetSlotForLabel()
+		if err != nil {
+			return 0, err
+		}
+	}
+
+	session, err := c.client.OpenSession(c.slot, pkcs11.CKF_SERIAL_SESSION|pkcs11.CKF_RW_SESSION)
+	if err != nil {
+		return 0, fmt.Errorf("failed to open session: %w", err)
+	}
+	err = c.client.Login(session, pkcs11.CKU_USER, c.pin)
+	if err != nil {
+		return 0, fmt.Errorf("failed to login: %w", err)
+	}
+	return session, nil
+}
+
+func (c *Pkcs11Client) CloseSession(session pkcs11.SessionHandle) {
+	if c.client == nil {
+		return
+	}
+	c.client.Logout(session)
+	c.client.CloseSession(session)
+}
+
+//
+func (c *Pkcs11Client) FindKey(session pkcs11.SessionHandle, key Pkcs11Key, typ uint) ([]pkcs11.ObjectHandle, error) {
+	template := []*pkcs11.Attribute{
+		pkcs11.NewAttribute(pkcs11.CKA_LABEL, []byte(key.label)),
+		pkcs11.NewAttribute(typ, true),
+	}
+	if keyIdBytes, err := hex.DecodeString(key.id); err == nil {
+		template = append(template, pkcs11.NewAttribute(pkcs11.CKA_ID, keyIdBytes))
+	}
+	if c.mechanism != 0 {
+		keyTypeString, err := GetKeyTypeFromMech(c.mechanism)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get key type from mechanism: %s", err)
+		}
+		template = append(template, pkcs11.NewAttribute(pkcs11.CKA_KEY_TYPE, keyTypeString))
+	}
+
+	if err := c.client.FindObjectsInit(session, template); err != nil {
+		return nil, fmt.Errorf("failed to pkcs11 FindObjectsInit: %s", err)
+	}
+	obj, _, err := c.client.FindObjects(session, 2)
```

Comment:
How do we want to handle multiple results? I think right now, given the above comment about uniqueness, we're at the mercy of the library as to what order it returns results in, which might be random.

Thoughts?

Reply:
There is a part in `Encrypt` and `Decrypt` that I'll move into `FindKey`: (and I will change the text too).
```
	if len(obj) != 1 {
		return nil, nil, fmt.Errorf("expected 1 object, got %d", len(obj))
	}
```
Since we're asking for 2 objects max, we can detect if there is more than 1 and then return an error.